scripts/generate_matrix.py

```diff
@@ -0,0 +1,227 @@
+# !/usr/bin/env python3
+"""
+Script to generate a test matrix of Python and package versions for CI testing.
+
+This script fetches the list of supported Python versions and the versions of a
+specified package available on PyPI, determines compatible combinations based
+on package 'requires_python' metadata, and outputs a JSON matrix suitable for
+use in CI workflows.
+
+Features:
+- Generalized to work with any package (not just NumPy).
+- Caches HTTP requests to reduce network calls.
+- Prevents code duplication by using reusable functions.
+- Includes thorough doctests and documentation.
+
+Usage:
+    python generate_matrix.py [PACKAGE_NAME]
+
+Example:
+    python generate_matrix.py numpy
+
+"""
+
+import dataclasses
+import importlib.metadata
+import json
+import sys
+import urllib.error
+import urllib.request
+from functools import lru_cache
+
+# Constants for caching
+CACHE_SIZE = 128  # Adjust as needed
+
+
+@dataclasses.dataclass(frozen=True, eq=True, order=True)
+class Version:
+    major: int
+    minor: int
+    micro: int
+
+    def __str__(self) -> str:
+        return f"{self.major}.{self.minor}.{self.micro}"
+
+    @property
+    def minor_version(self) -> tuple[int, int]:
+        return self.major, self.minor
+
+    @classmethod
+    def from_str(cls, version_str: str) -> "Version":
+        if version_str.count(".") == 1:
+            version_str += ".0"
+        return cls(*map(int, version_str.strip().split(".")))
+
+
+def get_minimum_python_version(package: str) -> Version:
+    """
+    Fetch the minimum Python version specified in pyproject.toml.
+
+    Returns:
+        str: The minimum required version of the package.
+
+    Examples:
+        >>> get_minimum_python_version()  # doctest: +ELLIPSIS
+        Version(major=3, ...)
+    """
+    raw_version = importlib.metadata.metadata(package)["Requires-Python"]
+    if "<" in raw_version:
+        raise NotImplementedError("Version specifier with upper bound not yet supported!")
+
+    return Version.from_str(raw_version.replace(">=", "").replace("~=", ""))
```

Comment:
This looks a lot like `scipy_minimum_python()` in `scripts/version_bounds.py`, so maybe this can be used there, or the other way around, to make it more DRY?

Reply:
I forgot to remove that file, it's obsolete since I've moved the functions here.